File: scripts/utils/check-capabilities.js
import { execFile } from 'child_process';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import { dirname, join, resolve, normalize } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Validate and sanitize path to prevent directory traversal
function validatePath(pathToCheck, baseDir) {
  const resolved = resolve(baseDir, pathToCheck);
  const normalized = normalize(resolved);
  if (!normalized.startsWith(normalize(baseDir))) {
    throw new Error(`Invalid path: ${pathToCheck} - path traversal detected`);
  }
  return normalized;
}

// Read MATLAB configuration with validation
const configPath = join(process.cwd(), 'config', 'matlab.config.json');
if (!existsSync(configPath)) {
  throw new Error(`Configuration file not found: ${configPath}`);
}
const config = JSON.parse(readFileSync(configPath, 'utf8'));

// Validate matlab path from config
if (!config.matlab || !config.matlab.installPath || typeof config.matlab.installPath !== 'string') {
  throw new Error('Invalid MATLAB configuration: installPath must be a string');
}

const matlabPath = validatePath(config.matlab.installPath, process.cwd());
const matlabExe = join(matlabPath, 'matlab.exe');
const scriptPath = join(__dirname, 'check_matlab_capabilities.m');

console.log('🔍 Checking MATLAB Capabilities...\n');
console.log(`MATLAB Path: ${matlabExe}\n`);

// Check if MATLAB executable exists
if (!existsSync(matlabExe)) {
  console.log('⚠️  MATLAB executable not found at the specified path.');
  console.log('Please update config/matlab.config.json with the correct MATLAB installation path.');
  console.log('Skipping MATLAB capabilities check.\n');
  process.exit(0);
}

// MATLAB command to run the script and exit
// -batch flag runs MATLAB in batch mode (non-interactive)
// -r runs the specified command/script
// Sanitize script path for MATLAB command
const sanitizedScriptPath = scriptPath.replace(/\\/g, '/').replace(/'/g, "''");

// Use execFile with array of arguments to prevent command injection
const matlabArgs = ['-batch', `run('${sanitizedScriptPath}')`];

execFile(
  matlabExe,
  matlabArgs,
  {
    maxBuffer: 10 * 1024 * 1024, // 10MB buffer for large outputs
    timeout: 60000, // 60 second timeout
  },
  (error, stdout, stderr) => {
    if (error) {
      console.error('❌ Error running MATLAB:');
      console.error(error.message);
      if (stderr) {
        console.error('STDERR:', stderr);
      }
      return;
    }

    // Display the output
    console.log(stdout);

    if (stderr && !stderr.includes('Warning')) {
      console.error('Warnings:', stderr);
    }

    // Save report to file
    const reportPath = join(__dirname, 'matlab-capabilities-report.txt');
    writeFileSync(reportPath, stdout, 'utf8');
    console.log(`\n✅ Report saved to: ${reportPath}`);
  }
);

File: scripts/utils/check-capabilities.py
#!/usr/bin/env python3
"""
Python script to check MATLAB capabilities using MATLAB Engine API
Alternative approach if MATLAB Engine for Python is installed
"""

import json
import os
import subprocess
import sys
from pathlib import Path


def check_matlab_via_cli():
    """Check MATLAB capabilities via command line interface"""
    # Read MATLAB configuration
    config_path = Path(__file__).parent / 'matlab.config.json'
    with open(config_path, 'r') as f:
        config = json.load(f)

    matlab_path = Path(config['matlab']['installPath'])
    matlab_exe = matlab_path / 'matlab.exe'
    script_path = Path(__file__).parent / 'check_matlab_capabilities.m'

    print('🔍 Checking MATLAB Capabilities...\n')
    print(f'MATLAB Path: {matlab_exe}\n')

    # Run MATLAB in batch mode
    cmd = [
        str(matlab_exe),
        '-batch',
        f"run('{script_path.as_posix()}')"
    ]

    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=60,
            cwd=Path(__file__).parent
        )

        print(result.stdout)

        if result.stderr:
            print('Warnings:', result.stderr, file=sys.stderr)

        # Save report
        report_path = Path(__file__).parent / 'matlab-capabilities-report.txt'
        with open(report_path, 'w', encoding='utf-8') as f:
            f.write(result.stdout)

        print(f'\n✅ Report saved to: {report_path}')

    except subprocess.TimeoutExpired:
        print('❌ MATLAB command timed out')
    except FileNotFoundError:
        print(f'❌ MATLAB executable not found at: {matlab_exe}')
    except Exception as e:
        print(f'❌ Error: {e}')

def check_matlab_via_engine():
    """Check MATLAB capabilities using MATLAB Engine API (if installed)"""
    try:
        import matlab.engine
        print('🔍 Checking MATLAB Capabilities via Engine API...\n')

        # Start MATLAB engine
        eng = matlab.engine.start_matlab()

        # Run the capabilities script
        script_path = Path(__file__).parent / 'check_matlab_capabilities.m'
        eng.run(str(script_path), nargout=0)

        # Get version info
        version = eng.version()
        print(f'MATLAB Version: {version}\n')

        # Get installed toolboxes
        toolboxes = eng.ver()
        print('Installed Toolboxes:')
        for i in range(len(toolboxes)):
            print(f'  - {toolboxes[i]["Name"]} (Version: {toolboxes[i]["Version"]})')

        eng.quit()

    except ImportError:
        print('⚠️  MATLAB Engine for Python not installed.')
        print('   Falling back to command-line interface...\n')
        check_matlab_via_cli()
    except Exception as e:
        print(f'❌ Error using MATLAB Engine: {e}')
        print('   Falling back to command-line interface...\n')
        check_matlab_via_cli()

if __name__ == '__main__':
    # Try Engine API first, fallback to CLI
    check_matlab_via_engine()
